### PR TITLE
Pr 6

### PR DESCRIPTION
**Pull request by laurynas-biveinis**
## Summary by CodeRabbit

* **New Features**
  * Added a tool to archive Org headline subtrees by headline or ID URI; returns archive file path and archived entry URI.

* **Documentation**
  * Added README and release notes describing usage, response fields, archive-location precedence, save/unsaved-change behavior, and examples.

* **Tests**
  * Expanded test suite for archiving success/failure, destination precedence, ID preservation, buffer/save guards, and related edge cases.

* **Refactor**
  * Improved multi-buffer file-saving and refresh behavior for reliable edits.

* **Chores**
  * Updated editor indentation settings for consistency.

### CODE REVIEW

**Comment by laurynas-biveinis** [comment]
(LLM agent)

Thanks for the reviews (both Claude rounds and CodeRabbit). All review fix-ups are squashed into the feature commit and the branch was force-pushed, so line numbers below refer to the current tip.

## Addressed

- In-file archiving now has a URI-resolvability test: the returned `org-id://` resolves via `resources/read` with **no** change to `org-mcp-allowed-files` (the entry stays in the already-allowed source file) — the contrast to the separate-file case.
- `cl-assert (buffer-file-name)` added to `org-mcp--complete-and-save` to make the "must be called inside a visiting buffer" invariant explicit.
- `by-id` test strengthened to match the full archive file via a complete-file regex that pins the preserved ID.
- Pinned `org-adapt-indentation` to `nil` in the archive setup macro (fixes the Emacs 27.2 CI failure — its default flipped in Org 9.5).
- Fixture-docstring rewordings, tightened edebug specs, extracted the duplicated pre-existing-entry fixture, full UUID in the README example, and the `org-archive--compute-location` comment now also notes the wrong-guard risk.

## Pushed back

**1. `(should (null org-id-locations))` is not fragile — it's load-bearing.** `org-mcp-test--with-id-tracking` `let`-binds `org-id-locations` to `nil`, and the no-leak test registers zero IDs, so it stays `nil` *unless* the guard fails to abort before `org-id-get-create` (which would populate it). `null` is therefore the precise regression detector; a `(hash-table-count …)` rewrite would obscure that. Kept the assertion, added a clarifying comment.

**2. The `%s` docstring correction was itself inaccurate.** `org-archive.el` (Emacs 30.2, `org-archive--compute-location`, line 200) expands `%s` via `file-name-nondirectory` — the file name *with* extension, minus directory — not "the full path stripped of its extension." Corrected the docstring to the actual semantics.

**3. `org-mcp--complete-and-save`'s docstring was not stale.** It already references `(buffer-file-name)` and carries no `file-path` parameter, so there was nothing to update there; added the `cl-assert` instead.

**4. CodeRabbit's suggested `by-id` patch would not have matched.** It reused `org-mcp-test--expected-archive-file-regex`, which is tailored to the simple fixture (`* TODO Task to Archive` / `Some content here.`). The `by-id` test archives `content-with-id-todo` (`* TODO Task with ID` / three content lines), so that regex never matches. Added a dedicated complete-file regex (`expected-archive-with-id-file-regex`) instead.

## Declined

- **`fboundp` guard around `org-archive--compute-location`:** the function exists in every supported Emacs (the package requires 27.1), byte-compilation already warns if it ever disappears, and the inline comment documents both the failure symptom and the local-replication fix. A runtime guard would be dead code under all supported configurations.
- **Pure-wording nits:** the `; blank lines` comment matches the existing convention in `expected-archive-file-regex`, and the explicit `save-buffer` is required for the `org-archive-subtree-save-file-p nil` path (harmless and idempotent otherwise), so neither was changed.

File: org-mcp.el [hunk 1]
```diff
@@ -1894,6 +1920,102 @@ MCP Parameters:
...
+              ;; accumulate buffers visiting archive files.  A buffer
+              ;; the user already had open is left untouched.  Guard on
+              ;; liveness too: a save hook may have killed it already,
+              ;; and `with-current-buffer' on a dead buffer would error.
+              (when (and archive-buffer
+                         (buffer-live-p archive-buffer)
+                         (not archive-buffer-existed))
+                (with-current-buffer archive-buffer
+                  (set-buffer-modified-p nil))
+                (kill-buffer archive-buffer))))
```
[the remark this answers is not in the document]
Comment (laurynas-biveinis) [reply org-mcp.el]: (LLM agent)

Respectfully, this is a false positive — the in-memory archive buffer is never the only copy, so killing it on the error path does not lose data.

The protected body's ordering is: (1) save the archive buffer, (2) refresh archive buffers, (3) `write-region` the now-empty source to disk, (4) refresh source buffers. The source **file on disk** is only emptied at step 3, which runs *after* the archive is saved at step 1:

- Error at/before step 1 (archive `save-buffer` fails): the source file on disk still holds the subtree (step 3 never ran), so killing the in-memory archive buffer loses nothing — the source file is intact and a retry works.
- Error after step 1: the archive file on disk holds the subtree, so killing the already-saved archive buffer is fine.

There is no window in which both disk files lack the subtree. The finding conflates the source *buffer* (which is unwound) with the source *file* (untouched until step 3). `org-mcp-test-archive-subtree-persists-archive-before-source` pins exactly this recovery: it stubs the step-3 `write-region` to fail and asserts the archive file has the entry and the source is unchanged.

The suggested `(unless (buffer-modified-p) (kill-buffer))` would instead leak a modified, tool-opened buffer on the save-failure path — contradicting the cleanup's purpose — for no data-safety gain. No change made.

File: org-mcp.el [hunk 2]
```diff
@@ -2319,6 +2441,29 @@ Returns JSON object:
...
+        Formats:
+          - org-headline://{absolute-path}#{url-encoded-path}
+          - org-id://{uuid}
+
+Returns JSON object:
+  success - Always true on success (boolean)
+  archive_file - Absolute path to the archive file (string)
+  uri - org-id:// URI of the archived headline (string);
+        resolvable via resources/read only if the archive file is
+        in org-mcp-allowed-files, otherwise informational only"
```
[the remark this answers is not in the document]
Comment (laurynas-biveinis) [reply org-mcp.el:2500]: (LLM agent)

Addressed, though not in the `:description`. The unsaved-buffer precondition is cross-cutting behavior, and org-mcp deliberately centralizes that in the server-level `org-mcp--instructions` constant (sent to clients via MCP `initialize`) rather than repeating it per tool — its docstring states it "holds the guidance that would otherwise be repeated in every tool and resource description," and no mutating tool duplicates it in its `:description`.

The actual gap was that `org-archive-subtree` was missing from that instructions text. Fixed by adding a sentence there covering its failure on unsaved changes in a buffer visiting **either** the source or the destination archive file (the dual-file aspect you noted).

I deliberately did not append it to the existing four-tool list in those instructions: those four also perform leading-header-block validation that `org-archive-subtree` does not, so listing it there would misstate its behavior.

**Comment by laurynas-biveinis** [comment]
(LLM agent)

Responses to the latest review round (`org-archive-subtree`):

**Done — `stringp` assert on `compute-location` (your "Moderate" item).** Added `(cl-assert (stringp …))`, inlined into the `archive-file` binding so it fires *before* `separate-archive`'s `string=` would otherwise mask a non-string with an opaque type error. This is distinct from the earlier `fboundp` suggestion I declined: byte-compilation already warns if the function disappears, but it cannot catch a return-*shape* change, so the assert earns its place and also documents the car-is-string assumption. It partially closes the "no test for unexpected shape" gap below.

**Done — test-cleanup helper.** Added `org-mcp-test--discard-buffer-visiting`. All four cleanup sites were the same "kill any buffer visiting FILE, discarding changes" unit, so a file-taking helper removes more duplication (the `find-buffer-visiting` + guard too) than a buffer-only one.

**Declined — replace `when separate-archive` with `when archive-buffer` (refresh guard).** Verified functionally equivalent: `archive-buffer` is nil exactly when no buffer visits the archive file, and `refresh-file-buffers` is then a no-op (a `string=` match implies a truename match, so `find-buffer-visiting` would have returned non-nil). But `when separate-archive` better expresses intent ("refresh the *other* buffers of the separate archive file") and is more robust if `archive-buffer`'s computation is ever narrowed — e.g. to tool-opened buffers only, after which `when archive-buffer` would silently skip refreshing a user's buffer. The swap trades that robustness for cosmetic symmetry, so I kept it.

**Declined — test for an unexpected `compute-location` return shape.** As you noted, this needs mocking Org internals; the new `cl-assert` documents the assumption instead, which is the next-best thing.

**Comment by laurynas-biveinis** [comment]
(LLM agent)

Responses to this round (`org-archive-subtree`):

**Done — guard-on-custom-location test (your "Medium" item).** Added `org-mcp-test-archive-subtree-guard-fires-on-custom-location`: it sets a per-headline `:ARCHIVE:` property redirecting to a `_custom` file, leaves a modified buffer visiting that file, and asserts the unsaved-changes guard fires on it (and the source is unchanged). That pins agreement between the tool's pre-flight `org-archive--compute-location` call and the file `org-archive-subtree` actually opens — exactly the location-mismatch drift you flagged. The pre-existing `honors-archive-property` covered only the success path on the custom location, so this is new coverage.

**Done — README example.** The request is now an `org-headline://` URI and the response returns a freshly minted `org-id://` UUID, so the example demonstrates the ID-minting behavior rather than the `org-id://` passthrough no-op.

**Done — comments on the three macro-skipping tests.** Added `;; No archive bindings needed: this aborts before any archive logic runs.` to `modified-buffer-error`, `non-string-uri`, and `nonexistent-headline`.

**Declined — coerce `archive-buffer-existed` to an explicit boolean.** Using a buffer object as a truthy predicate is idiomatic Elisp, and the `-existed` name already signals the intent; a forced `(… t)` coercion would add noise (and would have to retain the `separate-archive` gate the suggestion drops). Kept as-is.

**Comment by laurynas-biveinis** [comment]
(LLM agent)

Both nitpicks from the latest review are addressed (commit amended into the feature commit).

**Done — save-buffer failure cleanup test.** Added `org-mcp-test-archive-subtree-kills-opened-buffer-on-save-failure`: with `org-archive-subtree-save-file-p` nil (so the tool performs the only archive save), it stubs `save-buffer` to signal, then asserts the tool-opened archive buffer is still reclaimed, **no** archive file is written, and the **source file still holds the subtree** byte-for-byte. That locks down the step-1 failure path — and pins the data-safety property from the earlier discussion (the entry survives in the source file because the failure precedes emptying it).

**Done — single whole-file regex with the returned ID.** Converted the shared `org-mcp-test--expected-archive-file-regex` constant into an `org-mcp-test--archive-file-regex (&optional id)` helper. With an `id` it requires the `:ID:` line inside the PROPERTIES drawer, so `org-mcp-test-archive-subtree-success` now verifies the complete archive file in a single match (the separate `:ID:` substring check is removed). The two other callers pass no `id` and keep the generic match — DRY, and consistent with both the repo's single-complete-file-regex rule and the existing `org-mcp-test--expected-archive-with-id-file-regex` used by the by-id test.

314 tests pass; `./check.sh` is green.

**Comment by laurynas-biveinis** [comment]
(LLM agent)

Responses to this round (`org-archive-subtree`):

**Done — the two fixture-docstring CLAUDE.md fixes.** Both now describe data shape rather than operation/fixture context:

- `org-mcp-test--expected-archive-infile-source-regex`: "an Org file with a top-level `Archived Tasks' heading, under it a level-2 `TODO Task to Archive' entry with a PROPERTIES drawer, and a single body line" (dropped "after in-file archiving").
- `org-mcp-test--expected-archive-with-id-file-regex`: now describes the mode-line, section header, and the `Task with ID' TODO whose original `:ID:' is preserved among the ARCHIVE_* properties, followed by three body lines (dropped the `content-with-id-todo' reference).

**Declined — canary cross-reference in the `org-archive--compute-location` comment.** The comment is already long, and naming specific test names in a source comment invites rot. The `#+ARCHIVE:`/`:ARCHIVE:` precedence tests and the `org-mcp-test-archive-subtree-guard-fires-on-custom-location` test already serve as the drift canary without a code comment pointing at them.

**Declined — version ranges instead of "Org 9.5"/"Emacs 27.2" in the `with-archive-setup` docstring.** These are fixed historical facts, not stale-prone: the `org-adapt-indentation` default flipped in Org 9.5, and Emacs 27.2 is the specific CI version whose bundled Org exhibited the indentation that broke the tests. Naming them is more useful for a future reader than a vague range.

314 tests pass; `./check.sh` is green.

**Comment by laurynas-biveinis** [comment]
(LLM agent)

Respectfully declining both items in this round.

**Issue 1 — the quoted comment rule is not in CLAUDE.md.** The review attributes to CLAUDE.md: *"Never write multi-paragraph docstrings or multi-line comment blocks — one short line max."* That sentence does not appear in either the project `CLAUDE.md` or the global one. The actual "Code Comments" guidance says the opposite of a length cap:

- Do not add comments that restate the code; prefer "why" over "what";
- "Use comments to explain complex algorithms, non-obvious side effects, or historical context."

A private-API dependency (`org-archive--compute-location`) with two distinct failure modes, and the crash-safety ordering rationale, are exactly the non-obvious side effects / historical context the guideline endorses — and the review itself grants "the WHY is genuinely non-obvious… so the comments belong." Earlier review rounds also praised these same comments as "thorough and honest… exactly the right documentation per the project's comment conventions." Collapsing them to one line each on the basis of a rule that isn't in the repo would lose information the real guideline asks us to keep, so I'm keeping them as-is.

**Issue 2 — the fixture docstrings comply with the actual rule.** The project rule is "Test-fixture `defconst` docstrings describe the data and its shape or role, not the tests that use it." The `content-archive-custom-location` / `content-archive-keyword-location` docstrings name no test — they explain a non-obvious literal token in the data (`%s` is org-archive's own placeholder, not a Lisp `format` arg) and the fixture's role (which archive file the value resolves to). That is data/role description, not test rationale. A prior round explicitly found this docstring "adequate," and another had it corrected for *accuracy* (the `%s` expansion is `file-name-nondirectory`, extension kept). Stretching "not the tests" to "no runtime semantics" goes beyond what the rule says, so I'm leaving these as well.

No code changes this round.

File: org-mcp-test.el [hunk 3]
```diff
@@ -5945,6 +6080,623 @@ Some quote
...
+               (should (alist-get 'error response))))
+           (should (file-exists-p archive-file))
+           (should
+            (string-match-p
+             (org-mcp-test--archive-file-regex)
+             (org-mcp-test--read-file archive-file)))
+           (should
+            (string-match-p
+             "Task to Archive"
+             (org-mcp-test--read-file test-file))))
```
[the remark this answers is not in the document]
Comment (laurynas-biveinis) [reply org-mcp-test.el]: (LLM agent)

Fixed. `org-mcp-test-archive-subtree-persists-archive-before-source` now asserts the complete unchanged source image instead of the `"Task to Archive"` substring:

```elisp
(org-mcp-test--verify-file-matches
 test-file
 (concat "\\`"
         (regexp-quote org-mcp-test--content-archive-simple)
         "\\'"))
```

so a truncation or partial corruption of the source on the write-failure path now fails the test. For consistency I applied the same idiom to `org-mcp-test-archive-subtree-kills-opened-buffer-on-save-failure` (it had used `string=` for the same "source unchanged" assertion), so both write-failure tests pin the full file image the same way.

314 tests pass; `./check.sh` is green.

**Comment by laurynas-biveinis** [comment]
@coderabbitai review

**Comment by laurynas-biveinis** [comment]
@coderabbitai full review

File: org-mcp.el [hunk 4]
```diff
@@ -1894,6 +1925,110 @@ MCP Parameters:
...
+                  ;; rather than vanishing from both.
+                  ;; `org-archive-subtree' may not save the archive
+                  ;; itself (`org-archive-subtree-save-file-p'), so
+                  ;; save it explicitly.
+                  (when archive-buffer
+                    (with-current-buffer archive-buffer
+                      (save-buffer)))
+                  (write-region (point-min) (point-max) file-path)
+                  (org-mcp--refresh-file-buffers file-path
+                                                 (current-buffer)))
```
[the remark this answers is not in the document]
Comment (laurynas-biveinis) [reply org-mcp.el]: (LLM agent)

Declining this one — the suggested `(org-mcp--refresh-file-buffers archive-file archive-buffer)` is a guaranteed no-op, so it would re-add dead code (an earlier review already removed exactly this call from this function for the same reason).

Reasoning:

- `archive-buffer` is bound to `(find-buffer-visiting archive-file)` *after* `org-archive-subtree` ran. `org-archive-subtree` opens the archive via `find-file-noselect`, which reuses/creates exactly one buffer per file truename — Emacs does not allow two buffers to visit the same file. So the sole buffer visiting `archive-file` *is* `archive-buffer`.
- `org-mcp--refresh-file-buffers archive-file archive-buffer` iterates `(buffer-list)`, matches `(string= buffer-file-name archive-file)`, and excludes `(eq buf archive-buffer)`. The only candidate is the excluded buffer, so the loop reverts nothing.

The source/archive asymmetry is intentional, not an oversight: the source-side refresh is load-bearing precisely because the tool's own temp buffer reaches `file-path` via `set-visited-file-name` (in `org-mcp--with-visiting-org-file`), which bypasses `find-file-noselect`'s dedup — so a user's pre-existing source buffer can coexist with the tool's temp buffer and genuinely needs reverting. The archive side has no second buffer (`find-file-noselect` dedups), so there is nothing to refresh. Adding the call would also reintroduce a misleading "refresh other buffers" framing for code that refreshes none.

**Comment by laurynas-biveinis** [comment]
(LLM agent)

@coderabbitai Addressed the remaining nitpick in the latest push (commit `7191361`, amended into the feature commit).

Hoisted the inline anchored unchanged-source image out of the two write-failure tests into a shared `defconst`:

```elisp
(defconst org-mcp-test--expected-archive-simple-unchanged-source-regex
  (concat "\\`"
          (regexp-quote org-mcp-test--content-archive-simple)
          "\\'")
  "Regex matching the simple archive source left unchanged after a
failed archive -- the original fixture image, byte for byte.")
```

Both `org-mcp-test-archive-subtree-persists-archive-before-source` and `org-mcp-test-archive-subtree-kills-opened-buffer-on-save-failure` now reference it instead of rebuilding the `concat`/`regexp-quote` expression inline, matching the project's before/after-image `defconst` convention and the sibling source-regex constants.

316 tests pass; `./check.sh` is green.

The other comment (archive-file refresh) was declined as a guaranteed no-op — see that thread.

**Comment by laurynas-biveinis** [comment]
(LLM agent)

Thanks for the review. I went through all three issues against the Org 30.2 `org-archive.el` source; declining all three, with evidence.

**Issue 1 — preexisting archive buffer "left stale": false positive.**

The premise that the buffer "won't auto-reflect the newly added entry" doesn't hold. When the archive target is a separate file, `org-archive-subtree` obtains the destination buffer via `(find-file-noselect afile 'nowarn)` (`org-archive.el:257`), which *reuses* an already-open buffer, then `(set-buffer buffer)` (`org-archive.el:308`) and inserts the subtree **into that buffer in memory**. The tool then `save-buffer`s it. So a user's preexisting archive buffer already shows the new entry — buffer and disk are in sync; there is nothing stale to refresh.

The suggested `(org-mcp--refresh-file-buffers archive-file archive-buffer)` is also a no-op: after archiving, `archive-buffer` *is* that preexisting buffer (Emacs keeps one buffer per file truename), so passing it as `except-buffer` skips the sole visitor and reverts nothing. The `nil`-except variant would instead force-revert the user's own buffer, discarding point/undo/fold state to reload content it already holds — a regression, not a fix. (CodeRabbit raised the same suggestion and accepted this reasoning.)

**Issue 2 — strengthen the internal-function comment: declining.**

This comment was deliberately trimmed to its two load-bearing facts earlier in this PR (it had been flagged as oversized). Re-adding "tested against Org 9.x / failure-mode" narration regresses that and runs against the terse-comment guideline. The suggested wording is also inaccurate: the `condition-case` is there to translate a *malformed user-supplied location*; it would only incidentally catch a future `void-function` and would mislabel it as `"Invalid archive location …"`, so documenting it as a designed version-compat fallback overstates it.

The real underlying question — the declared `(emacs "27.1")` floor vs. the Org-9.4-only `org-archive--compute-location` — is legitimate, but it belongs to the package manifest / CI matrix, not this comment's prose. Tracking it separately.

**Issue 3 — docstring `%s` / "extension kept": non-issue.**

The docstring describes `%s` as "`file-name-nondirectory`, extension kept." `org-archive.el:200` expands it as `(format file-fmt (file-name-nondirectory current-file))` — extension kept. The docstring matches the implementation exactly; no change needed.

Net: no code changes from this review. Appreciate the thorough pass, especially confirming the save-ordering and buffer-ownership design.

**Comment by laurynas-biveinis** [comment]
(LLM agent) Thanks for the review. No correctness issues raised; declining all five items, with reasoning:

1. **`org-archive--compute-location` version note** — The existing comment (`org-mcp.el:1958-1965`) already flags it as Org-internal and load-bearing. A signature/semantics change would fail the archive tests loudly rather than silently; the minimum Org version belongs in package metadata, not a local comment. No change.

2. **`(debug t)` → structured `(debug (...))`** — The premise is incorrect: `(debug t)` is the dominant idiom in this file, not an outlier (`org-mcp-test.el:1864, 1890, 2004, 2034, 2054`). `org-mcp-test--assert-archive-error` is a thin wrapper; `(debug t)` is consistent. No change.

3. **`#+ARCHIVE:` keyword path comment** — The precedence (ARCHIVE property → `#+ARCHIVE:` → `org-archive-location`) is already documented in the tool description (`org-mcp.el:2470-2472`), README, and NEWS. Adding a local comment duplicates that. No change.

4. **In-file archive skips the archive-buffer block** — This would restate the code: `archive-buffer` is `nil` precisely when `separate-archive` is `nil`, and the `(when archive-buffer ...)` gate makes the fall-through self-evident. No change.

5. **`\\(?:\n\\)*` → `\\(?:\n\\)+`** — The `*` is correct and intentionally tolerant. Pinning Org's exact blank-line count for a structural separator (not a value under test) would make the test fragile across Org versions for no behavioral guarantee. No change.

File: org-mcp-test.el [hunk 5]
```diff
@@ -1672,6 +1821,60 @@ Very deep content."
...
+  "Bind ARCHIVE-FILE to PATH and evaluate BODY in an `unwind-protect'.
+On exit, kill any buffer visiting ARCHIVE-FILE (discarding unsaved
+changes) and delete the file if it exists."
+  (declare (indent 2) (debug (symbolp form body)))
+  `(let ((,archive-file ,path))
+     (unwind-protect
+         (progn ,@body)
+       (org-mcp-test--discard-buffer-visiting ,archive-file)
+       (when (file-exists-p ,archive-file)
+         (delete-file ,archive-file)))))
```
[the remark this answers is not in the document]
Comment (laurynas-biveinis) [reply org-mcp-test.el:1840]: (LLM agent) Applied in 21e4c29. `org-mcp-test--discard-buffer-visiting` now iterates `buffer-list` and kills every buffer visiting the file, matching `org-mcp-test--with-archive-file`'s documented "kill any buffer visiting ARCHIVE-FILE" contract. With the shared teardown now covering multiple visitors, `org-mcp-test-archive-subtree-refreshes-other-archive-buffers` dropped its bespoke `unwind-protect`/`dolist` cleanup. Full suite green (319/319).

File: org-mcp-test.el [hunk 6]
```diff
@@ -5945,6 +6206,474 @@ Some quote
...
+                       test-file)))
+      (cl-letf (((symbol-function 'org-mcp--refresh-file-buffers)
+                 (lambda (&rest _)
+                   (error "Simulated refresh failure"))))
+        (should
+         (alist-get 'error (org-mcp-test--archive-tool-response uri))))
+      (should-not (find-buffer-visiting default-archive-file))
+      (org-mcp-test--verify-file-matches
+       default-archive-file
+       org-mcp-test--expected-archive-simple-file-regex))))
```
[the remark this answers is not in the document]
Comment (laurynas-biveinis) [reply org-mcp-test.el:6600]: (LLM agent) Applied in 21e4c29. The stub in `org-mcp-test-archive-subtree-kills-opened-buffer-on-error` now signals only for the source file, so the archive refresh runs normally and the failure fires on the post-`write-region` source refresh — the path the docstring describes. Added an `org-mcp-test--expected-archive-source-regex` assertion confirming the source is emptied on disk, and tightened the docstring. Verified against the implementation: refresh order is archive (after `save-buffer`) → `write-region` source → source refresh. Full suite green (319/319).

**Comment by laurynas-biveinis** [comment]
(LLM agent) Thanks for the review. Addressed in 21e4c29:

- **Regex duplication** (`org-mcp-test--expected-archive-with-id-file-regex`): now calls `org-mcp-test--archive-file-regex` with the ID, heading, and a multi-line body. No helper extension was needed — the existing `body` parameter already runs through `regexp-quote`, and passing the multi-line literal yields a byte-identical regex (verified).
- **In-file `org-archive-subtree-save-file-p` path** and **`org-archive--compute-location` internal API**: acknowledged as no-change, per your notes.

The same commit also folds in two test-fidelity fixes from CodeRabbit's pass (the unconditional refresh-failure stub was tripping on the archive refresh before the source write; and `org-mcp-test--discard-buffer-visiting` only killed a single visitor).

**Comment by laurynas-biveinis** [comment]
(LLM agent) Thanks for the deeper pass. After verifying each item against the Org 9.7.11 source and runtime, I am declining all three.

**1. Pre-flight location resolution / `org-get-local-archive-location` (Medium) — declined, factually incorrect.**

`org-archive-subtree` does **not** resolve via `org-get-local-archive-location`; it uses the exact expression the tool copies (org-archive.el:248-250, Org 9.7.11):

```elisp
(org-archive--compute-location
 (or (org-entry-get nil "ARCHIVE" 'inherit)
     org-archive-location))
```

The pre-flight is byte-for-byte identical to it, so the unsaved-changes guard checks the same file `org-archive-subtree` will use — divergence is impossible by construction. Moreover `org-get-local-archive-location` is unbound in Org 9.7.11 (`fboundp` → nil; the only archive-location function present is `org-archive--compute-location`), so switching to it would be a `void-function` error. Even if it existed, it would *introduce* the very divergence the comment guards against, since it is not what `org-archive-subtree` calls. The `#+ARCHIVE:` keyword is honored today (proven by the passing `org-mcp-test-archive-subtree-honors-archive-keyword`) precisely because `org-entry-get ... 'inherit` picks it up — the same mechanism org-archive itself relies on.

**2. Extract `org-mcp--resolve-archive-file` helper (Low) — declined, YAGNI.**

There is a single call site, so the stated benefit (localising breakage to one place) is already met. A one-line, single-caller helper provides no localisation and does not break up any long nested block, so it fails the project's YAGNI policy.

**3. Stale `org-id-locations` after mid-archive throw (Low) — acknowledged, no change.**

Correct that a throw after `org-id-get-create` leaves a transient id→source entry whose UUID was never written to disk. org-id tolerates this by design: the next lookup rescans the file, does not find the id, and abandons the stale entry. The honest behaviour is therefore a self-healing transient, which cannot be asserted as "not polluted"; adding unwind-protect cleanup would be complexity for a recoverable edge case.

Net: no changes from this review.
